Fix for Order by issue in select into statement with identity function

deepakshi-mittal opened the pull request:
### Description
Fix for Order by issue in select into statement with identity function

Modified  transformSelectIntoStmt, remove unused variable queryString
Modified  SortClause and Distinct Clause of select into, removed identity function if exists in it
Added alter table add identity statement which will be executed after select into statement
Added IsA check for SortGroupClause
Removed unused variable tsql_select_into_seq_oid
Removed smallint and int functions,
Dropped function objects from upgrade scripts
Used strcasecmp inplace of strncasecmp

Added test cases for select into identity with 
- top clause 
- distinct clause
- order by clause

Task: BABEL-4388

Engine Pr:https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/247

### Issues Resolved

BABEL-4388, BABEL-539

### Test Scenarios Covered ###
* **Use case based -**

Added test cases for select into identity with 
- Different syntax IDENTITY(datatype) ,  IDENTITY(datatype, seed), IDENTITY(datatype, seed, increment)
- negative seed and increment values
- bigint, numeric, decimal datatypes etc
- top clause 
- distinct clause
- order by clause asc, desc
- order by multiple columns
- selecting into temporary table
- Self Join 

* **Negative based -**
- User defined datatypes (Not supported)
- Non numeric types like char (Not allowed)
- calling internal functions directly
- Two identity columns in a query
- IDENTITY() without specifying typename
-  IDENTITY(int) without specifying column name for it


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Fix for Order by issue with identity function

Added Alter table add identity after creating table

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Fixed targetlists entry number removing identity

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Fix for failing test case, removed unused hook

Adjusted Target List resno

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added hook to move identity function at the end of target list

Modified  transformSelectIntoStmt, remove unused variable queryString
Modified  SortClause and Distinct Clause of select into, removed
identity function if exists in it
Added alter table add identity statement which will be executed after
select into statement

Added test cases for select into identity with
- top clause
- distinct clause
- order by clause

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

xhfanhe commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4518,4 +4500,35 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
+		if (IsA(res->val, FuncCall))
+		{
+			FuncCall *fc = (FuncCall *)res->val;
+			if (strncasecmp(strVal(llast(fc->funcname)), "identity_into", strlen("identity_into")) == 0)
```
What if funcname is "identity_into_xxx", will this still pass? I think strcmp is sufficient enough.

deepakshi-mittal replied:
Changed to strcasecmp "identity_into_bigint", Removed two extra function int and smallint and just kep one, bigint which will take care of all numeric case

xhfanhe commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4518,4 +4500,35 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
+
+	foreach (o_target, targetlist)
+	{
+		ResTarget *res = (ResTarget *)lfirst(o_target);
```
Need to first check if this Node is a ResTarget

deepakshi-mittal replied:
Removed this hook as the code works without modifying target list

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -233,8 +233,35 @@ TsqlFunctionConvert(TypeName *typename, Node *arg, Node *style, bool try, int lo
 Node *
 TsqlFunctionIdentityInto(TypeName *typename, Node *seed, Node *increment, int location)
```
will identity function be called in this case ?

deepakshi-mittal replied:
The actual functions wont be invoked, but we need the typename, seed and increment value. I can make a single function

deepakshi-mittal replied:
Removed two extra function int and smallint and just kep one, bigint which will take care of all numeric cases

xhfanhe commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5743,33 +5744,64 @@ static List *transformSelectIntoStmt(CreateTableAsStmt *stmt, const char *queryS
+				ListCell *olitem;
+				foreach (olitem, q->sortClause)
+				{
+					SortGroupClause *sortcl = (SortGroupClause *)lfirst(olitem);
```
Always check IsA() to make sure we are getting the correct struct

deepakshi-mittal replied:
Done

xhfanhe commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5663,49 +5666,49 @@ static int64 get_identity_into_args(Node *node)
 			TargetEntry *tle = (TargetEntry *)lfirst(elements);
 
-			if (tle->expr && IsA(tle->expr, FuncExpr) && strncasecmp(get_func_name(((FuncExpr *)(tle->expr))->funcid), "identity_into", strlen( "identity_into")) == 0)
+			if (tle->expr && IsA(tle->expr, FuncExpr) && strncasecmp(get_func_name(((FuncExpr *)(tle->expr))->funcid), "identity_into", strlen("identity_into")) == 0)
```
Same doubt for strncasecmp here

deepakshi-mittal replied:
changed to strcasecmp

deepakshi-mittal committed:
Addressed Code review comments, Removed hook no more needed

Removed hook bbf_rewrite_targetlist_hook
Added IsA check for SortGroupClause
Removed unused variable tsql_select_into_seq_oid
Removed smallint and int functions,
Dropped function objects from upgrade scripts
Used strcasecmp inplace of strncasecmp

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Fix for failing upgrade test

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2312,34 +2311,22 @@ bbf_set_context_info(PG_FUNCTION_ARGS)
- * Common function to get next value of sequence for Identity function in Select Into
- */
-static int64 
-get_identity_next_value(void)
```
Its no longer needed as we are using alter table add identity statement now

deepakshi-mittal committed:
Added comment for deprecated functions

Task: BABEL-4388

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>
